src/lib/tls/tls13/tls_server_impl_13.cpp

```diff
@@ -0,0 +1,294 @@
+   auto server_cert_chain = client_hello.find_certificate_chain(credentials_manager());
+   BOTAN_ASSERT_NOMSG(!server_cert_chain.empty());
+
+   auto private_key = credentials_manager().private_key_for(server_cert_chain.front(), "tls-server",
+                      client_hello.sni_hostname());
+   BOTAN_ASSERT_NONNULL(private_key);
+
+   // TODO: ALPN - Invoke Callbacks::tls_server_choose_app_protocol() with
+   //       suggestions sent by the client. This might happen in the Encrypted
+   //       Extensions constructor. Also implement Channel::application_protocol().
+
+   aggregate_handshake_messages()
+   .add(m_handshake_state.sending(Encrypted_Extensions(client_hello, policy(), callbacks())))
+   .add(m_handshake_state.sending(Certificate_13(server_cert_chain, Connection_Side::SERVER, {}, callbacks())))
+   .add(m_handshake_state.sending(Certificate_Verify_13(client_hello.signature_schemes(), Connection_Side::SERVER,
+                                  *private_key, policy(), m_transcript_hash.current(), callbacks(), rng())))
+   .add(m_handshake_state.sending(Finished_13(m_cipher_state.get(), m_transcript_hash.current())))
+   .send();
+
+   m_cipher_state->advance_with_server_finished(m_transcript_hash.current());
+
+   // TODO: For Client Authentication this should expect appropriate client handshake messages
+   //       once we support/implement it.
+   m_transitions.set_expected_next(FINISHED);
+   }
+
+void Server_Impl_13::handle_reply_to_client_hello(const Hello_Retry_Request& hello_retry_request)
+   {
+   auto cipher = Ciphersuite::by_id(hello_retry_request.ciphersuite());
+   BOTAN_ASSERT_NOMSG(cipher.has_value());  // should work, since we chose that suite
+
+   m_transcript_hash = Transcript_Hash_State::recreate_after_hello_retry_request(cipher->prf_algo(), m_transcript_hash);
+
+   m_transitions.set_expected_next(CLIENT_HELLO);
+   }
+
+void Server_Impl_13::handle(const Client_Hello_12& ch)
+   {
+   // The detailed handling of the TLS 1.2 compliant Client Hello is left to
+   // the TLS 1.2 server implementation.
+   BOTAN_UNUSED(ch);
+
+   // RFC 8446 Appendix D.2
+   //    If the "supported_versions" extension is absent and the server only
+   //    supports versions greater than ClientHello.legacy_version, the server
+   //    MUST abort the handshake with a "protocol_version" alert.
+   //
+   // If we're not expecting a downgrade, we only support TLS 1.3.
+   if(!expects_downgrade())
+      {
+      throw TLS_Exception(Alert::PROTOCOL_VERSION, "Received a legacy Client Hello");
+      }
+
+   downgrade();
+   }
+
+void Server_Impl_13::handle(const Client_Hello_13& client_hello)
+   {
+   const auto& exts = client_hello.extensions();
+
+   const bool is_initial_client_hello = !m_handshake_state.has_hello_retry_request();
+
+   if(is_initial_client_hello)
+      {
+      const auto preferred_version = client_hello.highest_supported_version(policy());
+      if(!preferred_version)
```

Comment:
Frankly, I don't remember that we discussed that before. Anyway, this might actually point to an implementation issue. So far, the second Client Hello is not compared to the original one. I.e. the server _does not_ ensure that the Client Hello changed only in the aspects outlined by [RFC 8446 4.1.2](https://www.rfc-editor.org/rfc/rfc8446#section-4.1.2).

Curiously, BoGo tests don't seem to cover this. I'll need to look into it. Good catch!

Reply:
https://github.com/randombit/botan/pull/3053/commits/ca25dda541ebcbae61d902aeea41f992630ca3df contains additional validation of the updated Client Hello. It does not specifically validate the contents of extensions, though. I'm frankly not sure whether that is even needed.

Regarding the 'supported version' changing after the Hello Retry Request: If the object is a `Client_Hello_13`, it's highest supported version is definitely TLS 1.3 (otherwise the parser would have classified it as `Client_Hello_12` and we would have run into the check added in the same commit: https://github.com/randombit/botan/pull/3053/commits/ca25dda541ebcbae61d902aeea41f992630ca3df#diff-fcf2471ec5c28b66eae46f45d0e53de60b2afa2fe6f94d219038bc58c9d635be.